bftengine/src/bftengine/ControlStateManager.cpp

```diff
@@ -87,4 +90,61 @@ void ControlStateManager::clearCheckpointToStopAt() {
   saveReservedPage(0, data.size(), data.data());
 }
 
+std::pair<bool, std::string> ControlStateManager::canUnwedge() {
+  std::optional<std::string> result;
+  if (!bftEngine::IControlHandler::instance()->isOnNOutOfNCheckpoint()) {
+    return {false, "Replica has not reached the wedge point yet"};
+  }
+
+  std::string replica_id = std::to_string(ReplicaConfig::instance().getreplicaId());
+  auto sigManager = impl::SigManager::instance();
+  std::string sig(sigManager->getMySigLength(), '\0');
+  sigManager->sign(replica_id.c_str(), replica_id.size(), sig.data(), sig.size());
```

Comment:
This is not a good practice in my opinion: The signature will always look the same. I am not sure how many times this signature will occur, but there should be a "salt". I would recommend adding a counter (wedge/unwedge cycle) so that eavesdrops attacker see a different signature every time.
Sign: replica id + cycle number. I think it would be more secure.

Reply:
Well, I am aware we are using open source and the attacker maybe read the code. But in general - having access to the exact data that was signed can open the get to a sophisticated attacker to get the signature private key.
So it is better to use "salt" before signing.

Reply:
@glevkovich  What do you think about using the last sequence number as salt. It is guaranteed to be the same in all replicas since they are wedged.

Reply:
good idea!